lua/dap.lua

```diff
@@ -343,6 +328,52 @@ function Session.event_output(_, body)
   repl.append(body.output, '$')
 end
 
+function Session:_current_frame_index()
+  local current_frame_index = nil
+  if self.stopped_thread_id and self.threads[self.stopped_thread_id].frames then
+    local frames = self.threads[self.stopped_thread_id].frames
+    for i, f in ipairs(frames) do
+      if self.current_frame.id == f.id then
+        current_frame_index = i
+      end
+    end
+  end
+  return current_frame_index
+end
+
+function Session:_request_scopes(current_frame)
+  self:request('scopes', { frameId = current_frame.id }, function(_, scopes_resp)
+    if not scopes_resp or not scopes_resp.scopes then return end
+
+    current_frame.scopes = {}
+    for _, scope in pairs(scopes_resp.scopes) do
+
+      table.insert(current_frame.scopes, scope)
+      if not scope.expensive then
+        self:request('variables', { variablesReference = scope.variablesReference }, function(_, variables_resp)
+          if not variables_resp then return end
+
+          scope.variables = variables_resp.variables
+        end)
+      end
+    end
+  end)
+end
+
+function Session:_frame_delta(delta)
+  if self.stopped_thread_id and self.threads[self.stopped_thread_id].frames then
+    local frames = self.threads[self.stopped_thread_id].frames
+    local current_frame_index = self:_current_frame_index()
+
+    current_frame_index = current_frame_index + delta
+    if 0 < current_frame_index and current_frame_index <= #frames then
+      self.current_frame = frames[current_frame_index]
```

Comment:
Do you think it would make sense to wrap around? Or would that be confusing?
Or maybe print a message when hitting the top-most or bottom-most frame?

Reply:
I guess this would be enough feedback:

dap> .up

  a()
  b()
  c()
  d()
 → e()

dap> .up

  a()
  b()
  c()
  d()
 → e()

Reply:
Yes I agree. Want to include it? Could also be a follow up - I think this PR is good to go

Reply:
It's included.